aten/src/ATen/native/sparse/SparseBinaryOpIntersectionKernel.cpp

```diff
@@ -15,18 +16,82 @@ struct CPUKernelLauncher {
   }
 };
 
-
 struct MulOp {
-  static Tensor apply(const Tensor& a, const Tensor& b) {
-    return a.mul(b);
+  template <typename scalar_t>
+  static scalar_t apply(scalar_t a, scalar_t b) {
+    return a * b;
+  }
+};
+
+template <>
+bool MulOp::apply(bool a, bool b) {
+  return a && b;
+}
+
+template <typename binary_op_t>
+struct CPUValueSelectionIntersectionKernel {
+  static Tensor apply(
+      const Tensor& lhs_values,
+      const Tensor& lhs_select_idx,
+      const Tensor& rhs_values,
+      const Tensor& rhs_select_idx) {
+    auto iter = make_value_selection_intersection_iter(
+        lhs_values,
+        lhs_select_idx,
+        rhs_values,
+        rhs_select_idx);
+    auto res_values = iter.tensor(0);
+
+    const auto lhs_nnz_stride = lhs_values.stride(0);
+    const auto rhs_nnz_stride = rhs_values.stride(0);
+
+    AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND3(
+        ScalarType::Bool, ScalarType::Half, ScalarType::BFloat16, res_values.scalar_type(),
+        "binary_op_intersection_cpu", [&] {
+          AT_DISPATCH_INDEX_TYPES(lhs_select_idx.scalar_type(),
+              "binary_op_intersection_cpu", [&] {
+                auto loop = [&](char** data, const int64_t* strides, int64_t n) {
+                  auto* ptr_res_values_bytes = data[0];
+                  const auto* ptr_lhs_values_bytes = data[1];
+                  const auto* ptr_lhs_select_idx_bytes = data[2];
+                  const auto* ptr_rhs_values_bytes = data[3];
+                  const auto* ptr_rhs_select_idx_bytes = data[4];
+
+                  for (int64_t i = 0; i < n; ++i) {
+                    // Exctract data
+                    auto* RESTRICT ptr_res_values = reinterpret_cast<scalar_t*>(ptr_res_values_bytes);
+                    const auto* ptr_lhs_values = reinterpret_cast<const scalar_t*>(ptr_lhs_values_bytes);
+                    const auto lhs_nnz_idx = *reinterpret_cast<const index_t*>(ptr_lhs_select_idx_bytes);
+                    const auto* ptr_rhs_values = reinterpret_cast<const scalar_t*>(ptr_rhs_values_bytes);
+                    const auto rhs_nnz_idx = *reinterpret_cast<const index_t*>(ptr_rhs_select_idx_bytes);
```

Comment:
@amjames , could you please confirm whether skipping `restrict` for these value pointers is justified?